File: leetcode/python/best_time_to_buy_and_sell_stock_121.py
import pytest
from typing import List

testCases = [
    {"name": "case0", "input": [7, 1, 5, 3, 6, 4], "want": 5},
    {
        "name": "case1",
        "input": [7, 6, 4, 3, 1],
        "want": 0,
    },
    {"name": "high low high", "input": [7, 3, 2, 5, 8], "want": 6},
    {"name": "low high low", "input": [3, 4, 8, 5, 1], "want": 5},
]

"""
You are given an array prices where prices[i] is the price of a given stock on the ith day.

You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.

Ref: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/ 
"""


class Solution:
    # input: [1,2,3,4,5]
    def maxProfitOn2(self, prices: List[int]) -> int:
        maxProfit = 0
        #
        for i in range(0, len(prices) - 1):
            # invariant: 0 < i < j < len(prices)
            # i = 0, prices[i] = 1
            for j in range(i + 1, len(prices)):
                # j = 1, prices[i] = 2
                curProfit = prices[j] - prices[i]
                # curProfit = 2 - 1 = 1
                if curProfit > maxProfit:
                    # 2 - 1 = 1
                    maxProfit = curProfit
                    # update maxProfit = 1
        return maxProfit

    def maxProfitOn(self, prices: List[int]) -> int:
        maxProfit = 0
        buyIn = prices[0]
        curProfit = 0
        for i in range(1, len(prices)):
            curProfit = prices[i] - buyIn
            if curProfit < 0:
                buyIn = prices[i]
            maxProfit = max(maxProfit, curProfit)
        return maxProfit


@pytest.mark.parametrize("testCase", testCases, ids=lambda testCase: testCase["name"])
def test_maxProfitOn2(testCase):
    sol = Solution()
    input = testCase["input"]
    want = testCase["want"]
    got = sol.maxProfitOn2(input)
    assert got == want


@pytest.mark.parametrize("testCase", testCases, ids=lambda testCase: testCase["name"])
def test_maxProfitOn(testCase):
    sol = Solution()
    input = testCase["input"]
    want = testCase["want"]
    got = sol.maxProfitOn(input)
    assert got == want

File: leetcode/java/app/src/main/java/com/leetcode/BestTimeToBuyAndSellStock121.java
package com.leetcode;

import com.leetcode.BinaryTree.TreeNode;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;
import java.util.function.BiConsumer;


/**
 * 给定一个数组 prices ，它的第 i 个元素 prices[i] 表示一支给定股票第 i 天的价格。
 * <p>
 * 你只能选择 某一天 买入这只股票，并选择在 未来的某一个不同的日子 卖出该股票。设计一个算法来计算你所能获取的最大利润。
 * <p>
 * 返回你可以从这笔交易中获取的最大利润。如果你不能获取任何利润，返回 0 。
 * 提示：
 *
 * 1 <= prices.length <= 105
 * 0 <= prices[i] <= 104
 * <a href="https://leetcode.cn/problems/best-time-to-buy-and-sell-stock/description/">...</a>
 */
public class BestTimeToBuyAndSellStock121 {
    static class Solution {
        public int maxProfit(int[] prices) {
            // FIXME: java assert
            // { 2, 7, 3, 1, 10, 7 } , want: 9
            assert prices.length > 0;

            int buyInPrice = prices[0];
            int maxProfit = Integer.MIN_VALUE;
            for (int i = 1; i < prices.length; i++) {
                int currentPrice = prices[i];
                maxProfit = Math.max(maxProfit, currentPrice - buyInPrice);
                if (currentPrice - buyInPrice < 0) {
                    // buyIn
                    buyInPrice = currentPrice;
                }
            }

            return Math.max(maxProfit, 0);


            // Def of maxProfit:
            // 1. maxProfit = sellOutPrice - buyInPrice
            // 2. maxProfit > 0
            // buyIn price should be lowest.
            // sell price should be highest and sellOutDay > buyInDay
            // maxProfit > 0
            // when to update buyInPrice ?
            //   when we find a buyInPrice2 that buyInPrice2 < buyInPrice
            // when to update sellOutPrice ?
            //   when we find a buyInPrice2 that buyInPrice2 < buyInPrice
        }
    }
}
